eureka-client/src/main/java/com/netflix/discovery/shared/transport/decorator/RetryableEurekaHttpClient.java

```diff
@@ -152,10 +159,14 @@ public void shutdown() {
         List<EurekaEndpoint> candidateHosts = clusterResolver.getClusterEndpoints();
         quarantineSet.retainAll(candidateHosts);
 
-        // If all hosts are bad, we have no choice but start over again
-        if (quarantineSet.size() == candidateHosts.size()) {
+        // If enough hosts are bad, we have no choice but start over again
+        int threshold = candidateHosts.size() * transportConfig.getRetryableClientQuarantineRefreshPercentage() / 100;
+        if (quarantineSet.isEmpty()) {
+            // no-op
+        } else if (quarantineSet.size() >= threshold) {
```

Comment:
I am not sure if this is good approach. If first 50% servers are down, and threshold is 50%, the client will never be able to connect to any healthy instance, as quarantine set will be just before the first live instance reset to zero.